### ISSUE TITLE
How to use model to predict for a plain sentences  the semantic role labels ?

### ISSUE THREAD

**Issue by rematchka** [phueb/DLSRL#11]
Hi Philip,
i was wondering on how to use trained model to predict semantic role labels for plain input text?
thank you.

**Comment by phueb** [comment]
Train a model, then load it, and use the evaluate function in train_utils.py to get predictions for plain text sentences. Keep in mind you have to label the verb for which you want to generate SR labels in your plain text sentences.

**Comment by rematchka** [comment]
and where exactly do i have to label the verb? do i need to tokenize sentence?

**Comment by phueb** [comment]
Take a look at `data/conll05.dev.txt`. You'll see there is an integer before each sentence. This integer represents the position of the verb in the sentence (index starts at zero).

**Comment by rematchka** [comment]
do i need to pre-process plain text? i mean i just need to add index of verb i want SR to generate label for and give it to evaluate function in train_utils.py  or the data given to this function needed to be in special formate?

**Comment by phueb** [comment]
Take a look at `data/conll05.dev.txt` to understand the formatting. Notice that after the sentence there are three pipes which indicate the start of the SR labels. The `evaluate` function expects this format, but feel free to rewrite it not to expect SR labels.

**Comment by rematchka** [comment]
i meant now that i read raw sentence ex " The FCC effort collapsed . " should i call `evaluate` function directly on that sentence or should i call `get_propositions_from_file` as in `get_data` then 

```
  sentece_id = [words_to_ids(p[0], word_dict, True, word2embed) for p in raw_sent_props]
    sent_predicate_ids = get_predicate_ids(raw_sent_props, config)
    sent_label_ids = [words_to_ids(p[2], label_dict) for p in raw_sent_props]
 sent_data = [sentece_id, sent_predicate_ids, sent_label_ids] 
```
then call evaluate function using this list

**Comment by phueb** [comment]
`evaluate` requires that you pass a data object, a model, a tensorflow session, and some integer representing the training step. Therefore you need to invoke `get_data` to create a data object, using the path to your text files (train, dev, test) as one of the arguments, before calling `evaluate`. Then, inside `evaluate`, you will find:
```
x1, x2, y = shuffle_stack_pad(data, batch_size=batch_size, shuffle=False)
feed_dict = make_feed_dict(x1, x2, y, model, keep_prob=1.0)

# export confusion matrix
summary = sess.run(model.cm_summary, feed_dict=feed_dict)
model.train_writer.add_summary(summary, global_step)

# get predictions 
pred, gold = sess.run([model.nonzero_predicted_label_ids, model.nonzero_label_ids_flat], feed_dict=feed_dict)
```

Because you don't have gold SR labels, you can modify the last line to:
`pred = sess.run([model.nonzero_predicted_label_ids], feed_dict=feed_dict)`

Also, notice that `get_data` expects text files with SR labels. If you don't have any, the easiest way to circumvent this issue is to provide dummy SR labels. Keep in mind though, that this will invalidate the f1 score, if calculated using your dummy SR labels

**Comment by rematchka** [comment]
So devData will be pain input text and i need to add to it dummy SRL labels? `epoch, global_step` arguments in `evaluate` should i give them random numbers? after prediction i will get 
```
role
ARG0
ARG1 
ARG2
```

**Comment by phueb** [comment]
`pred` would be SR label ids which would need to be converted to labels using the `label_dict` that is returned by `get_data`. label_dict takes as input a label and returns a label id, so I would create the reverse dictionary. Please feel free to drop arguments like epoch and `global_step` from the `evaluate` function. I used those for printing information how far along training is.

**Comment by rematchka** [comment]
okay , now DEV_DATA_PATH (Plain text) will be argument to `get_data` using it's output `devData` which will be used in `evaluate` function to get predictions. what i don't get how i will now which word it has this id  and it's role   [role,ARG0,ARG1,ARG2]. why will i need `'label_dict'` as it contain dummy labels since i don't know real labels or did i get it wrong or does` label_dict` contain id : role irrespective to my data  ,but if saying that how to reverse map to get the word that have this id and role,but when i looked up label_dic is dictionary type that have  `self.str2idx , self.idx2str` is that what will i use to get label given id ,but how to get original word?

**Comment by phueb** [comment]
First, thank you for reminding me that `label_dict` is not a builtin dict object mapping labels to their ids. Like you said, it is an object which contains two dictionaries: `str2idx` maps labels to ids and `idx2str` maps ids to labels. So, no need to reverse a dictionary. Just do:
`predicted_sr_label = label_dict.idx2str[predicted_label_id]`

Second, you are concerned about the dummy labels. Good point. But this will not be an issue, because `label_dict` does not include labels from the dev text file (which will be dummies). This is due to:
`label_dict.accept_new = False`
This line occurs after `train_data` is created but before `dev_data` is created, and ensures that SR labels from your dev text file are not included as labels.

**Comment by rematchka** [comment]
what are expected SRL labels i will get is it [role,ARG0,ARG1,ARG2] or [A0,V,A1,A2]as stated in the paper or will i get a  BIO sequence?
` for i, j in zip(gold[:100], pred[:100]):
        print('a="{}", p="{}"'.format(i, j))`
so this basically this loops on the predicted SRL lables for each word in the sentence ,right? all i need to do is to discard the Unknown label if found for word ? 
if i have multiple verbs in a sentence as in case of  complex and compound sentences, do i have to state sentence twice in the file and only change the index of verb? or the model works only in the case of simple sentence (one verb only in sentence)?

**Comment by phueb** [comment]
Yes, please state the sentence twice in the dev text file - this is how it is done in the training text file. The SR labels are the same as in the paper, and use BIO tagging. Something like ARG0-B, ARG0-I, if I remember correctly.

**Comment by rematchka** [comment]
i was wondering does the model stuck at some value of f1 score . cause i have this problem that the model is stuck at  `f1 0.59 `most of the training time and loss doesn't decrease dramatically. another thing i kinda get this output from prediction `a="1",p="1"` while training is it normal?

**Comment by phueb** [comment]
What data is used to calculate the F1? Remember, if you are calculating it on your own dev data which has dummy labels, this would make sense. And can you be more specific about the output? Perhaps copy a larger portion of the output, highlighting the portion you are curious about?

**Comment by rematchka** [comment]
Actually i was training the model  from the folder data you uploaded on GitHub not on my own dev-data/Train Data . I calculate `F1 `after each epoch but i got stuck at `0.59` after `60 `epoch from start of training using `dev data` provided in this repo .

when evaluating model after each epoch in training phase that's the output i get
`for i, j in zip(gold[:100], pred[:100]): print('a="{}", p="{}"'.format(i, j))`
  `a="3", p="1"
a="6", p="1"
a="6", p="1"
a="6", p="1"
a="20", p="1"
a="1", p="1"
a="2", p="2"
a="10", p="3"
a="11", p="6"
a="11", p="6"
a="11", p="6"
a="14", p="6"
a="15", p="6"
a="1", p="6"
a="24", p="6"
a="25", p="1"
a="25", p="1"
a="25", p="1"
a="25", p="1"
a="25", p="1"
a="25", p="1"
a="25", p="1"
a="25", p="1"
a="25", p="1"
a="1", p="1"
a="3", p="1"
a="6", p="1"
a="6", p="1"
a="1", p="1"
a="2", p="2"
a="30", p="3"
a="31", p="6"
a="31", p="6"
a="31", p="6"
a="31", p="6"
a="31", p="6"
a="31", p="6"
a="31", p="6"
a="31", p="6"
a="31", p="1"
a="31", p="1"
a="31", p="1"
a="31", p="1"
a="1", p="1"
a="24", p="1"
a="25", p="1"
a="25", p="1"
a="25", p="1"
a="25", p="1"
a="25", p="1"
a="25", p="1"
a="25", p="1"
a="25", p="1"
a="25", p="1"
a="25", p="1"
a="25", p="1"
a="25", p="1"
a="25", p="1"
a="25", p="1"
a="25", p="1"
a="1", p="1"
a="3", p="1"
a="6", p="1"
a="6", p="1"
a="1", p="1"
a="1", p="1"
a="1", p="1"
a="2", p="2"
a="16", p="3"
a="17", p="6"
a="17", p="6"
a="17", p="6"
a="17", p="6"
a="12", p="6"
a="18", p="6"
a="18", p="6"
a="18", p="6"
a="18", p="1"
a="18", p="1"
a="1", p="1"
a="1", p="1"
a="1", p="1"
a="1", p="1"
a="1", p="1"
a="1", p="1"
a="1", p="1"
a="1", p="1"
a="1", p="1"
a="1", p="1"
a="1", p="1"
a="1", p="1"
a="1", p="1"
a="1", p="1"
a="1", p="1"
a="1", p="1"
a="1", p="1"
a="1", p="1"
a="3", p="1"
a="6", p="1"
a="6", p="1"`

**Comment by phueb** [comment]
I remember I was able to get an F1 score very similar to the original implementation - not sure why you are stuck. 
As for the output, I remember now: "a" is short for "actual" (or "gold") and "p" is short for "predicted. `gold` contains ids corresponding to the correct/actual SR labels, and `pred` contains the ids of the labels predicted by the model. This allows some very basic comparison. Looks like the model isn't doing so good in your example.

**Comment by rematchka** [comment]
I am right now at 360 epoch should i continue training or abort it since result of F1 is the same?
Would this problem due to  tensor-flow version (F1 being stuck at some value)?

**Comment by phueb** [comment]
It should be steadily increasing - no need to keep training it. Make sure you are using the "interleaved" (not "stacked") LSTM architecture. To do this, verify that in `__init__.py`, 
```
config_str = '''
{
  "train_batch_size" : 128,
  "architecture" : "interleaved", 
  "use_se_marker": false,
  "embed_size" : "100",
  "cell_size": 256,
  "num_layers" : 8,
  "max_grad_norm": 1.0,
  "keep_prob" : 0.9,
  "max_epochs": 500,
  "learning_rate": 1.0,
  "epsilon": 1e-6
}
'''
```

**Comment by rematchka** [comment]
is this a right graph for the model? 
![graph_run=](https://user-images.githubusercontent.com/25964922/54220502-9b580980-4502-11e9-82ba-55d735efabfa.png)

**Comment by phueb** [comment]
That graph is using the correct "interleaved" architecture, so that's not a problem. I can't think of any other reason you are getting bad performance. I will try to run the model on my machine and see if I can get better results, but I may not be able to do so until after next week.

**Comment by rematchka** [comment]
Thank you so much.

**Comment by phueb** [comment]
What version of tensorflow were you using when you ran the model?

**Comment by rematchka** [comment]
tensorflow 1.12

**Comment by phueb** [comment]
I am encountering the same problem you did. Since I developed the repository using tensorflow 1.4, I wonder whether it has to do with having since upgraded to 1.12 as you have.

**Comment by rematchka** [comment]
Should i try downgrading tensorflow version and rerun the training?

**Comment by phueb** [comment]
I think that would be the logical next step. Unfortunately I can't do this without going through much trouble on my machine, because of my CUDA installation.

**Comment by rematchka** [comment]
i have inquiry related to cuda downgrading version . if i have this message does that mean i cannot install this version?
![Screenshot (1)](https://user-images.githubusercontent.com/25964922/55029277-c53d2000-501a-11e9-8287-f50b4e538c26.png)

**Comment by phueb** [comment]
This is a question for a different thread. Keep in mind you can always try, and then reinstall what you had previously.

**Comment by rematchka** [comment]
i have inquiry regarding loss and f1 accuracy does  it change within 100 epoch or fluctuate? i mean in first 100 epochs  (0-100 epochs )should  f1 fluctuate around 59- 58 then the next 100 epochs (100-200 epochs) fluctuate around 68-69 and so on or should it increase gradually in first 100 epochs (0-100 epochs) to find different values for f1 score other than 59 & 58 ?

**Comment by phueb** [comment]
I remember it growing gradually when I developed the model. Did you get f1 past 0.59?

**Comment by rematchka** [comment]
i'm still at 30 epoch. but the model f1 score is still stuck at 0.59.

**Comment by rematchka** [comment]
I haven't passed 0.59 f1 score, but i think it might be due to this warning 

`UserWarning: Converting sparse IndexedSlices to a dense Tensor of unknown shape. This may consume a large amount of memory.
  "Converting sparse IndexedSlices to a dense Tensor of unknown shape. "`

**Comment by phueb** [comment]
What line in the code raises this error? Is this error generated using tensorflow 1.12?

**Comment by rematchka** [comment]
This is warning, generated in tensorflow versions 1.12 and 1.3 and 1.4.
when i checked it has to do with gradients.
i don't know which line generates it,but it's most probably theses lines
 ` gradients, variables = zip(*optimizer.compute_gradients(self.nonzero_mean_loss))
`
 `  gradients, _ = tf.clip_by_global_norm(gradients, config.max_grad_norm)`
`  self.update = optimizer.apply_gradients(zip(gradients, variables), name='update')`

i'm not sure though.

**Comment by phueb** [comment]
Why do you think that converting a sparse tensor to a dense tensor might impair the training of the model?

**Comment by rematchka** [comment]
1- when i was trying to to load model to continue training, i had an error `Unsuccessful TensorSliceReader constructor` which i thought it's somehow related to this warning.
2-  `UserWarning: Converting sparse IndexedSlices to a dense Tensor of unknown shape. This may consume a large amount of memory. "Converting sparse IndexedSlices to a dense Tensor of unknown shape. "`    when i was searching about this warning , i found some people saying that it make model crash and didn't continue training , and that this warning may cause slow model learning. Although other people said to safely to ignore the warning and it won't make any problem. 
i'm not sure also if it's causing problem to the bad model behavior or not, or that's due to another unseen problem.

i have inquiry, why isn't LSTM cell t initialized using orthogonal initialize, to be 
`cell = HighwayWrapper(DropoutWrapper(LSTMCell(size,initializer=tf.orthogonal_initializer()),
                                                 variational_recurrent=True,
                                                 dtype=tf.float32,
                                                 state_keep_prob=keep_prob))`
instead of `cell = HighwayWrapper(DropoutWrapper(LSTMCell(size),
                                                 variational_recurrent=True,
                                                 dtype=tf.float32,
                                                 state_keep_prob=keep_prob))`

**Comment by phueb** [comment]
I have ported the code to tensorflow 2.0.0b1. The model now works correctly as indicated by the f1 score evaluation (which calls the official evaluation script provided by the conll-05 shared task)
